modules/oe_webtools_media/tests/src/FunctionalJavascript/MediaSourceWebtoolsTest.php

```diff
@@ -102,6 +107,59 @@ public function testMediaWebtoolsSource(): void {
     }
   }
 
+  /**
+   * Tests the webtools media source support for wcloud.
+   */
+  public function testMediaWebtoolsSourceWcloud(): void {
+
+    $session = $this->getSession();
+    $page = $session->getPage();
+    $assert_session = $this->assertSession();
+
+    // Create a webtools media type for charts.
+    $this->createWebtoolsMediaType('test_media_webtools_type', 'chart');
+
+    // Create a webtools wcloud media item without an url.
+    $this->drupalGet("media/add/test_media_webtools_type");
+    $name = "Valid webtools wcloud item";
+    $assert_session->fieldExists('Name')->setValue($name);
+    $assert_session->fieldExists("Webtools Chart snippet")->setValue('{"utility": "wcloud"}');
+    $page->pressButton('Save');
+    $assert_session->pageTextContains("The provided wcloud url is not valid.");
+
+    // Create a webtools wcloud media item with an invalid url.
+    $assert_session->fieldExists("Webtools Chart snippet")->setValue('{"utility": "wcloud", "url": "not-a-url"}');
+    $page->pressButton('Save');
+    $assert_session->pageTextContains("The provided wcloud url is not valid.");
+
+    // Create a webtools wcloud media item with an url
+    // outside the europa.eu domain.
+    $assert_session->fieldExists("Webtools Chart snippet")->setValue('{"utility": "wcloud", "url": "https://google.com"}');
+    $page->pressButton('Save');
+    $assert_session->pageTextContains("The wcloud url needs to be in the europa.eu domain.");
+
+    // Create a webtools wcloud media item with an url that
+    // is not accessible/does not exist.
+    $assert_session->fieldExists("Webtools Chart snippet")->setValue('{"utility": "wcloud", "url": "https://europa.eu/fake-url"}');
+    $page->pressButton('Save');
+    $assert_session->pageTextContains("Could not parse contents of the wcloud url.");
+
+    // Create a webtools wcloud media item with an url that returns
+    // a correct response, but for the wrong widget type.
+    $assert_session->fieldExists("Webtools Chart snippet")->setValue('{"utility": "wcloud", "url": "https://europa.eu/correct-wcloud?widget=map"}');
+    $page->pressButton('Save');
+    $assert_session->pageTextContains("Invalid webtools chart snippet.");
+
+    // Create a webtools wcloud media item with an url that returns
+    // a correct response.
+    $assert_session->fieldExists("Webtools Chart snippet")->setValue('{"utility": "wcloud", "url": "https://europa.eu/correct-wcloud?widget=chart"}');
```

Comment:
Can we test with a valid url that indeed the snippet is resolved on render from the remote server? Or assert the response from the MockWcloudMiddleware?

Reply:
I don't think is needed. We are already asserting that we validate the contents of the url using the same validation as regular snippets and that is the only thing this module cares about since it only provides the media source.